setup.py

```diff
@@ -30,44 +16,138 @@ def PyPiDescription():
-    try:
-        # Check if CUDA is installed
-        cuda = ctypes.cdll.LoadLibrary('libcudart.so')
-        deviceCount = ctypes.c_int()
-        # get the number of supported GpUs
-        cuda.cudaGetDeviceCount(ctypes.byref(deviceCount))
-        if deviceCount.value > 0:
-            gpu_platform = True
-    except Exception as e:
-        gpu_platform = False
-if gpu_platform:
-    requirements = ['mxnet-cu90mkl>=1.2'] + requirements
-else:
-    requirements = ['mxnet-mkl>=1.2'] + requirements
-setup(
-    name='mxnet-model-server',
-    version=__version__.strip(),
-    description='Model Server for Apache MXNet is a tool for serving neural net models for inference',
-    long_description=PyPiDescription(),
-    url='https://github.com/awslabs/mxnet-model-server',
-    keywords='MXNet Model Server Serving Deep Learning Inference AI',
-    packages=pkgs,
-    install_requires=requirements,
-    entry_points={
-        'console_scripts': ['mxnet-model-server=mms.mxnet_model_server:start_serving',
-                            'mxnet-model-export=mms.export_model:export']
-    },
-    include_package_data=True,
-    license='Apache License Version 2.0'
-)
+
+import platform
+import os
+import ctypes
+import errno
+import time
+import sys
+from setuptools import setup, find_packages, Command
+import setuptools.command.build_py
+from shutil import copyfile
+import subprocess
+
+pkgs = find_packages(exclude=['*client_sdk_generator*', '*export_model*', '*swagger*', '*tests*'])
+source_server_file = os.path.abspath('frontend/server/build/libs/server-1.0.jar')
+dest_file_name = os.path.abspath('mms/frontend/model-server.jar')
+
+
+def pypi_description():
+    """Imports the long description for the project page"""
+    with open('PyPiDescription.rst') as df:
+        return df.read()
+
+
+def detect_model_server_version():
+    with open(os.path.abspath(os.path.join("mms", "version.py")), 'r') as vf:
+        exec(vf.read(), None, globals())
+    return __version__
+
+
+def select_mxnet(requirements=[]):
+    gpu_platform = False
+    if platform.system().lower() == 'linux':
+        try:
+            # Check if CUDA is installed
+            cuda = ctypes.cdll.LoadLibrary('libcudart.so')
+            device_count = ctypes.c_int()
+            # get the number of supported GpUs
+            cuda.cudaGetDeviceCount(ctypes.byref(device_count))
+            if device_count.value > 0:
+                gpu_platform = True
+        except Exception as e:
+            gpu_platform = False
+    if gpu_platform:
+        req = ['mxnet-cu90mkl>=1.2'] + requirements
+    else:
+        req = ['mxnet-mkl>=1.2'] + requirements
+
+    return req
+
+
+class BuildFrontEnd(Command):
+    description = 'Build Model Server Frontend'
+    user_options = []
+
+    def initialize_options(self):
+        pass
+
+    def finalize_options(self):
+        pass
+
+    def run(self):
+        try:
+            front_end_bin_dir = os.path.abspath('.') + '/mms/frontend'
```

Comment:
os.path.abspath('.') might throw exception, and will cause crash in exception handling code, due to front_end_bin_dir has not been initialized.

Reply:
The assumption with setup.py is we run python build in the same directory where setup.py exists. If there is an exception, it is raised to the above level and the build for setup.py will fail.